### PR TITLE
Fixes #1266: modifies sanitize.js to allow Telescope to parse img tags with data uris

### LINKED ISSUE #1266: Missing src attribute on img elements in posts when using data URIs
I notice a post today that is missing images.  The `` element is there, but it has no `src` attribute:

Screen Shot 2020-11-01 at 6 15 15 PM

The original post does have them: https://palak-chawla.blogspot.com/2020/10/release-02-pr-4.html, but they are using Data URIs:

Screen Shot 2020-11-01 at 6 17 40 PM

### PR DESCRIPTION
**Pull request by chrispinkney**
## Issue This PR Addresses
Fixes #1266 

## Type of Change

- [X] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
  * Modified `sanitize.js` to allow for `img` tags to use `data` uris as `src` so they can be successfully parsed and indexed by Telescope.
  * Adds a test to ensure `data` uris are sanitized properly.
  * Adds two more `http` and `https` schema tests for `img` tag with a data uri `src`.
  * Adds `blockquote` `test` and `cite` sanitation.

Hey everyone!

As per issue #1266, this PR allows images which have a `src` `data` uri type to be properly indexed. It passes all the tests on my end thus showing that data uris are being allowed by Telescope.

I tried to ensure that this works as expected by purging my redis db via `docker exec -it telescope_redis_1 redis-cli flushall` but it seems that once a post is indexed, it is how it is and forever shall be the way it was initially indexed. So instead of going back in time I disallowed all html attributes and waited for a post to come in. Once one arrived I noticed that it had no tags at all, just plaintext, so I reverted back to normal and allowed `img`s with a data uri.

While this isn't exactly a perfect sign that everything works as expected, given that `npm test sanitize-html` runs and returns that all tests pass, and specifically: `` returned as expected, I feel that Telescope should now properly parse `img`s with a data uri.

Let me know, thanks!

## Checklist

- [X] **Quality**: This PR builds and passes our npm test and works locally
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by chrispinkney** [comment]
So this is my first solo rebase that I did because I had like 3 commits. I also pushed them by accident, so I `git rebase HEAD~3 -i` and squashed the other two, then force pushed this to my branch and made a PR. I think I did all the steps right so let me know if I didn't!

**Comment by humphd** [comment]
@chrispinkney can you fix the title of this PR?  `fixes issue-1266` is not useful, as it doesn't reference what you're working on (the issue number is good, but describe the work too).

File: src/backend/utils/html/sanitize.js [hunk 1]
```diff
@@ -41,10 +41,12 @@ module.exports = function (dirty) {
       'tr',
       'ul',
     ],
+    allowedSchemesByTag: { img: ['http', 'https', 'data'] },
     allowedAttributes: {
       iframe: ['src'],
       img: ['src'],
       a: ['href'],
+      blockquote: ['cite'],
```
Comment (humphd) [inline src/backend/utils/html/sanitize.js:49]: Why is this here?

File: test/sanitize-html.test.js [hunk 2]
```diff
@@ -6,6 +6,30 @@ describe('Sanitize HTML', () => {
...
+  });
+
+  test('img links over https should be accepted (i.e. not sanitized)', () => {
+    const data = sanitizeHTML(
+      '<p><a><img src="https://www.telescope.com/https_image.jpg" /></a></p>'
+    );
+    expect(data).toBe('<p><a><img src="https://www.telescope.com/https_image.jpg" /></a></p>');
+  });
+
+  test('img links over http should be accepted (i.e. not sanitized)', () => {
```
Comment (humphd) [inline test/sanitize-html.test.js:26]: This is an interesting one, since Chrome is moving to block mixed content.  We may need to undo this later, since it will affect our security in the browser (we load over https but mixed content with http images will break us).

File: test/sanitize-html.test.js [hunk 3]
```diff
@@ -69,4 +93,13 @@ describe('Sanitize HTML', () => {
       '<table><tbody><tr><td><a href="www.senecacollege.ca"><img src="https://1.bp.blogspot.com/11.JPG" /></a></td></tr><tr><td>The Final Product</td></tr></tbody></table>'
     );
   });
+
+  test('<blockquote> should return properly and retain its citation', () => {
```
Comment (humphd) [inline test/sanitize-html.test.js:97]: This needs to be its own issue/pr

**Review by humphd — CHANGES_REQUESTED** [review]
Can you also add a test for protocol-less URLs? (i.e., ``)

File: src/backend/utils/html/sanitize.js (hunk 1 above)
Comment (chrispinkney) [reply src/backend/utils/html/sanitize.js:49]: Just wanted to add an additional test, hm, it does look irrelevant though in hindsight. I'll remove it from this PR.

File: test/sanitize-html.test.js (hunk 2 above)
Comment (chrispinkney) [reply test/sanitize-html.test.js:26]: Neat, should I add a comment specifying this?

File: src/backend/utils/html/sanitize.js (hunk 1 above)
Comment (humphd) [reply src/backend/utils/html/sanitize.js:49]: It's a great change, just do it in its own issue/pr.

File: test/sanitize-html.test.js (hunk 2 above)
Comment (humphd) [reply test/sanitize-html.test.js:26]: Sure, maybe link to https://web.dev/what-is-mixed-content/

**Comment by chrispinkney** [comment]
> 
> 
> Can you also add a test for protocol-less URLs? (i.e., ``)

I was going to add that too but there's already one in place [that does the same thing here](https://github.com/Seneca-CDOT/telescope/blob/b4151142d8dbda3f09343b91b072efb7003ca892/test/sanitize-html.test.js#L33).

**Comment by chrispinkney** [comment]
> 
> 
> @chrispinkney can you fix the title of this PR? `fixes issue-1266` is not useful, as it doesn't reference what you're working on (the issue number is good, but describe the work too).

Is this better?

**Review by humphd — APPROVED** [review]

**Comment by tonyvugithub** [comment]
@chrispinkney @humphd I wonder why the github action for macos failed. Should we retry with another push with a combined commit and see if all are passed? Right now one check has failed. 

This is the error in the macos build
`[HPM] Proxy rewrite rule created: "^/admin/build/log" ~> "/log"
[ 2020-11-03 23:27:02.081 ] ERROR (3824 on Mac-1604444996117.local): Something went wrong with search indexing`

**Comment by humphd** [comment]
That failure is not actually causing the tests to fail (i.e, the test expects the failure).  What's causing the failure is some process hanging and blocking the tests from exiting correctly (see #1284):

```
A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks.

Test Suites: 26 passed, 26 total
Tests:       169 passed, 169 total
Snapshots:   0 total
Time:        15.689 s
Ran all test suites.
Error: The operation was canceled.
```

**Comment by tonyvugithub** [comment]
@humphd I noticed that too. So I guess this PR is good to go ahead then if thats the case

**Review by tonyvugithub — APPROVED** [review]
